_docs-sources/foundations/iac-foundations/labels-tags.md

```diff
@@ -0,0 +1,264 @@
+
+If you are using custom tags for your resources for the purposes of tracking cost and usage, you will need to activate the tags in the Billing and Cost Management console in AWS. For more information see [User-defined cost allocation tags](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/custom-tags.html) and [Activating user-defined cost allocation tags](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/activating-tags.html).
+
+## Adding Labels and Tags
+
+Labels and tags are both specified as code and deployed with your resources. In this section, you will learn how to create a label as well as apply global and environment/resource-specific tags to your AWS resources.
+
+### Adding labels
+
+The simplest way to create a label is using a `locals` block and combining the recommended attributes into a single attribute. For example, if you were creating an AWS S3 bucket using a Gruntwork module you could set up the following locals block.
+
+```hcl
+terraform {
+  source = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/data-stores/s3-bucket?ref=v0.104.15"
+}
+
+include "root" {
+  path = find_in_parent_folders()
+}
+
+locals {
+    org = "gw"
+    env = "prd"
+    app = "coolapp"
+    identifier = "s3"
+    region = "uw2"
+
+    name = "${local.org}-${local.env}-${local.app}-${local.identifier}-${local.region}"
+}
+
+inputs = {
+  primary_bucket = local.name
+}
+```
+
+All Gruntwork modules expose the ability to name a resource, for which you can reference the attribute created in the `locals` block. If you are developing your own modules, you will need to ensure that you are exposing the ability to pass a name into the module.
+
+
+### Global tags
+
+Tags can be enforced globally by specifying them in the provider block. When specifying tags in the provider, the tags are automatically added to all resources created using that provider. Note that tags set at the provider level can be overridden, which is covered in [overriding tags](#overriding-tags), allowing you to specify all required tags in the provider with default values, then consumers of the provider to override the values as needed.
+
+#### AWS provider
+
+The AWS Provider supports the [default_tags](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#default_tags) parameter. This works with all AWS resources except `aws_autoscaling_group`. Here's an example:
+
+```hcl
+provider "aws" {
+  default_tags {
+    tags = {
+      Environment = "test"
+      Name        = "example"
+    }
+  }
+}
+```
+
+#### Terragrunt provider
+
+Terragrunt supports a [generate](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#generate) block. Instead of hard-coding the provider block in every single module, you can ensure that every repo that uses Terragrunt has a single, central generate block used by everything in that repo. That generate block can include the default_tags you want. For example:
```

Comment:
```suggestion
Terragrunt supports a [`generate`](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#generate) block. Instead of hard-coding the `provider` block in every single module, you can ensure that every repo that uses Terragrunt has a single, central `generate` block used by everything in that repo. That `generate` block can include the `default_tags` you want. For example:
```